src/hotspot/share/opto/parse2.cpp

```diff
@@ -76,57 +76,64 @@ void Parse::array_load(BasicType bt) {
   Node* adr = array_addressing(bt, 0, elemtype);
   if (stopped())  return;     // guaranteed null or range check
 
-  Node* idx = pop();
-  Node* ary = pop();
+  Node* array_index = pop();
+  Node* array = pop();
 
   // Handle inline type arrays
-  const TypeOopPtr* elemptr = elemtype->make_oopptr();
-  const TypeAryPtr* ary_t = _gvn.type(ary)->is_aryptr();
-  if (ary_t->is_flat()) {
+  const TypeOopPtr* element_ptr = elemtype->make_oopptr();
+  const TypeAryPtr* array_type = _gvn.type(array)->is_aryptr();
+  if (array_type->is_flat()) {
     // Load from flat inline type array
-    Node* vt = InlineTypeNode::make_from_flat(this, elemtype->inline_klass(), ary, adr);
-    push(vt);
+    Node* inline_type;
+    if (element_ptr->klass_is_exact()) {
+      inline_type = InlineTypeNode::make_from_flat(this, elemtype->inline_klass(), array, adr);
+    } else {
+      // Element type of flat array is not exact. Therefore, we cannot determine the flat array layout statically.
+      // Emit a runtime call to load the element from the flat array.
+      inline_type = load_from_unknown_flat_array(array, array_index, element_ptr);
+      inline_type = record_profile_for_speculation_at_array_load(inline_type);
```

Comment:
Cover missing case, other changes in `array_load()` are just renaming/refactoring